src/core/layout/qgslayoutexporter.cpp

```diff
@@ -43,26 +57,693 @@ void QgsLayoutExporter::renderPage( QPainter *painter, int page )
+    QHash< QGraphicsItem *, QGraphicsItem::CacheMode > mPrevCacheMode;
+};
+
+///@endcond PRIVATE
+
+void QgsLayoutExporter::renderRegion( QPainter *painter, const QRectF &region ) const
 {
   QPaintDevice *paintDevice = painter->device();
   if ( !paintDevice || !mLayout )
   {
     return;
   }
 
-  mLayout->context().mIsPreviewRender = false;
+  LayoutItemCacheSettingRestorer cacheRestorer( mLayout );
+  ( void )cacheRestorer;
 
 #if 0 //TODO
   setSnapLinesVisible( false );
 #endif
 
+  painter->setRenderHint( QPainter::Antialiasing, mLayout->context().flags() & QgsLayoutContext::FlagAntialiasing );
+
   mLayout->render( painter, QRectF( 0, 0, paintDevice->width(), paintDevice->height() ), region );
 
 #if 0 // TODO
   setSnapLinesVisible( true );
 #endif
+}
+
+QImage QgsLayoutExporter::renderRegionToImage( const QRectF &region, QSize imageSize, double dpi ) const
+{
+  double resolution = mLayout->context().dpi();
+  double oneInchInLayoutUnits = mLayout->convertToLayoutUnits( QgsLayoutMeasurement( 1, QgsUnitTypes::LayoutInches ) );
+  if ( imageSize.isValid() )
+  {
+    //output size in pixels specified, calculate resolution using average of
+    //derived x/y dpi
+    resolution = ( imageSize.width() / region.width()
+                   + imageSize.height() / region.height() ) / 2.0 * oneInchInLayoutUnits;
+  }
+  else if ( dpi > 0 )
+  {
+    //dpi overridden by function parameters
+    resolution = dpi;
+  }
+
+  int width = imageSize.isValid() ? imageSize.width()
+              : static_cast< int >( resolution * region.width() / oneInchInLayoutUnits );
+  int height = imageSize.isValid() ? imageSize.height()
+               : static_cast< int >( resolution * region.height() / oneInchInLayoutUnits );
+
+  QImage image( QSize( width, height ), QImage::Format_ARGB32 );
+  if ( !image.isNull() )
+  {
+    image.setDotsPerMeterX( resolution / 25.4 * 1000 );
+    image.setDotsPerMeterY( resolution / 25.4 * 1000 );
+    image.fill( Qt::transparent );
+    QPainter imagePainter( &image );
+    renderRegion( &imagePainter, region );
+    if ( !imagePainter.isActive() )
+      return QImage();
+  }
+
+  return image;
+}
+
+///@cond PRIVATE
+class LayoutContextSettingsRestorer
+{
+  public:
+
+    LayoutContextSettingsRestorer( QgsLayout *layout )
+      : mLayout( layout )
+      , mPreviousDpi( layout->context().dpi() )
+      , mPreviousFlags( layout->context().flags() )
+    {
+    }
+
+    ~LayoutContextSettingsRestorer()
+    {
+      mLayout->context().setDpi( mPreviousDpi );
+      mLayout->context().setFlags( mPreviousFlags );
+    }
+
+  private:
+    QgsLayout *mLayout = nullptr;
+    double mPreviousDpi = 0;
+    QgsLayoutContext::Flags mPreviousFlags = 0;
+};
+///@endcond PRIVATE
+
+QgsLayoutExporter::ExportResult QgsLayoutExporter::exportToImage( const QString &filePath, const QgsLayoutExporter::ImageExportSettings &s )
+{
+  ImageExportSettings settings = s;
+  if ( settings.dpi <= 0 )
+    settings.dpi = mLayout->context().dpi();
```

Comment:
I think that `mLayout` may be null at this point.

Reply:
BTW I think it may be the case in other methods too.

Reply:
This is a valid point for QgsLayoutExporter. I'll make the constructor require a reference instead of the pointer to avoid this.

Reply:
Fixed